### PR TITLE
Add compatibility with Docker dns servers

### LINKED ISSUE #1239: Add Monolith Discoverer that works for any dns server
### Is your feature request related to a problem? Please describe.

Setting up lots of monoliths and balancers locally for testing takes a lot of manual work. docker-compose has a feature where you can scale services:

```bash
docker-compose -f docker/with-balancer.docker-compose.yml up --scale opentogethertube=100 -d
```

### Describe the solution you'd like

Docker provides a dns server to allow apps to do service discovery, similar to Fly. See https://stackoverflow.com/questions/37683508/how-does-service-discovery-work-with-modern-docker-docker-compose

Copy `FlyMonolithDiscoverer` and adapt it to work with docker's dns server.

### Additional context

- Note that the docker containers don't have ipv6 addresses, so you'll have to query the dns server for ipv4 addresses instead.
- You'll also probably need to modify `deploy/ott-balancer-docker.toml` depending on what you name the config options

### PR DESCRIPTION
**Pull request by moreno-michael**
resolves #1239﻿

### CODE REVIEW

File: crates/ott-balancer/src/discovery/dns.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+#[async_trait]
+impl MonolithDiscoverer for DockerMonolithDiscoverer {
+    async fn discover(&mut self) -> anyhow::Result<Vec<MonolithConnectionConfig>> {
+        let resolver =
+            TokioAsyncResolver::tokio_from_system_conf().expect("failed to create resolver");
+
+        let docker_dns_server = format!("{.docker}")
+        let name = Name::from_str(&docker_service_name)?;
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:37]: This doesn't appear to do anything.

Also, the dns server is already determined from the system configuration. No need to specify it here.

File: crates/ott-balancer/src/discovery/dns.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use tracing::info;
+use trust_dns_resolver::TokioAsyncResolver;
+
+use super::*;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DockerDiscoveryConfig {
+    /// The port that monoliths should be listening on for load balancer connections.
+    pub monolith_port: u16,
+    pub fly_app: String,
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:11]: fly_app is not relevant here. The name of the monolith container is, however.

**Review by dyc3 — COMMENTED** [review]
Please fix the pr title, and don't forget to update the discovery diagrams before tomorrow.

File: crates/ott-balancer/src/discovery/dns.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#[derive(Debug, Clone, Deserialize)]
+pub struct DockerDiscoveryConfig {
+    /// The port that monoliths should be listening on for load balancer connections.
+    pub monolith_port: u16,
+    pub dns_server: Option<String>,
+}
+
+pub struct DockerMonolithDiscoverer {
+    config: DockerDiscoveryConfig,
+    query: Option<String>,
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:16]: This needs be configurable. move this to the config above and make it required.

File: crates/ott-balancer/src/discovery/dns.rs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+use async_trait::async_trait;
+use tracing::info;
+use trust_dns_resolver::TokioAsyncResolver;
+
+use super::*;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DockerDiscoveryConfig {
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:8]:
```suggestion
pub struct DnsDiscoveryConfig {
```

File: crates/ott-balancer/src/discovery/dns.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    /// The port that monoliths should be listening on for load balancer connections.
+    pub monolith_port: u16,
+    pub dns_server: Option<String>,
+}
+
+pub struct DockerMonolithDiscoverer {
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:14]:
```suggestion
pub struct DnsMonolithDiscoverer {
```

**Review by dyc3 — CHANGES_REQUESTED** [review]
Took a closer look, and this module isn't being included. add `mod dns` to `discovery.rs`, update `DiscoveryConfig` and fix the compiler errors. 

You should be able to run the balancer with the following configuration:

```toml
[discovery]
method = "dns"
query = "foo"
```

I think The demo we did during today's meeting worked because I had the balancer configured with the `manual` discoverer. That or I forgot to rebuild the container

File: crates/ott-balancer/src/discovery/dns.rs (hunk 3 above)
Comment (moreno-michael) [reply crates/ott-balancer/src/discovery/dns.rs:16]: So move just the query, right?

Comment (dyc3) [reply crates/ott-balancer/src/discovery/dns.rs:16]: yes, and make it not `Option`

File: crates/ott-balancer/src/discovery/dns.rs [hunk 6]
```diff
@@ -19,10 +19,10 @@ pub struct DnsMonolithDiscoverer {
 impl DnsMonolithDiscoverer {
     pub fn new(config: DnsDiscoveryConfig) -> Self {
         info!(
-            "Creating DockerMonolithDiscoverer, Docker DNS server: {}",
-            dns_server
+            "Creating DockerMonolithDiscoverer, Docker DNS server: {:?}",
+            config.dns_server
         );
-        let query = format!("{}", &config.dns_server);
+        let query = format!("{}", config.dns_server.clone().unwrap_or_default());
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:25]:
```suggestion
```

File: crates/ott-balancer/src/discovery/dns.rs [hunk 7]
```diff
@@ -19,10 +19,10 @@ pub struct DnsMonolithDiscoverer {
 impl DnsMonolithDiscoverer {
     pub fn new(config: DnsDiscoveryConfig) -> Self {
         info!(
-            "Creating DockerMonolithDiscoverer, Docker DNS server: {}",
-            dns_server
+            "Creating DockerMonolithDiscoverer, Docker DNS server: {:?}",
+            config.dns_server
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:23]:
```suggestion
            "Creating DnsMonolithDiscoverer, DNS server: {:?}",
            config.dns_server
```

File: crates/ott-balancer/src/discovery/dns.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        let query = format!("{}", config.dns_server.clone().unwrap_or_default());
+        Self { config }
+    }
+}
+
+#[async_trait]
+impl MonolithDiscoverer for DnsMonolithDiscoverer {
+    async fn discover(&mut self) -> anyhow::Result<Vec<MonolithConnectionConfig>> {
+        let resolver =
+            TokioAsyncResolver::tokio_from_system_conf().expect("failed to create resolver");
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:34]: The case where `self.config.dns_server` is `Some` is not taken into account here.

File: crates/ott-balancer/src/discovery/dns.rs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+use trust_dns_resolver::TokioAsyncResolver;
+
+use super::*;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DnsDiscoveryConfig {
+    /// The port that monoliths should be listening on for load balancer connections.
+    pub monolith_port: u16,
+    pub dns_server: Option<String>,
+    pub query: String
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:12]: Would be nice to have doc strings for these.
```suggestion
    /// The DNS server to query. Optional. If not provided, the system configuration will be used instead.
    pub dns_server: Option<String>,
    /// The A record to query. If using docker-compose, this should be the service name for the monolith.
    pub query: String
```

File: crates/ott-balancer/src/discovery/dns.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
+use async_trait::async_trait;
+use tracing::info;
+use trust_dns_resolver::TokioAsyncResolver;
+
+use super::*;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DnsDiscoveryConfig {
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:8]: This still hasn't been added to the `enum DiscoveryConfig` in `config.rs`.

**Review by dyc3 — CHANGES_REQUESTED** [review]
Per my previous review:

You should be able to run the balancer with the following configuration:

```toml
[discovery]
method = "dns"
query = "foo"
```

File: crates/ott-balancer/src/discovery/dns.rs (hunk 8 above)
Comment (moreno-michael) [reply crates/ott-balancer/src/discovery/dns.rs:35]: this comment will be resolved in Issue #1337

**Comment by moreno-michael** [comment]
@coderabbitai resolve the remaining comments

**Comment by dyc3** [comment]
No *you* need resolve the remaining comments. the bot can open an issue for you.

File: crates/ott-balancer/src/discovery/dns.rs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+pub struct DnsMonolithDiscoverer {
+    config: DnsDiscoveryConfig,
+}
+
+impl DnsMonolithDiscoverer {
+    pub fn new(config: DnsDiscoveryConfig) -> Self {
+        info!(
+            "Creating DockerMonolithDiscoverer, DNS server: {:?}",
```
Comment (dyc3) [inline crates/ott-balancer/src/discovery/dns.rs:24]:
```suggestion
            "Creating DnsMonolithDiscoverer, DNS server: {:?}",
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

**Review by dyc3 — APPROVED** [review]

**Pull request merged**